<launch>
  <master auto="start"/>
  <param name="/use_sim_time" value="false"/>
  <arg name="use_map_topic" default="true"/>
  <arg name="scan_topic" default="base_scan"/>
  <arg name="no_static_map" default="false"/>
  <arg name="base_global_planner" default="navfn/NavfnROS"/>
  <arg name="base_local_planner" default="dwa_local_planner/DWAPlannerROS"/>
  <arg name="map" value="$(find goto_odroid)/config/map/UTS_map.yaml"/>

  <!-- Run the map server -->
  <node pkg="map_server" type="map_server" name="map_server" args="$(arg map)"/>

  <!-- Reset USB connections -->
  <node pkg="goto_odroid" type="reset_wheelchair_.py" name="reset_wheelchair_connections"/>

  <!-- Run arduino serrial Uno extensional module -->
  <node pkg="rosserial_python" type="serial_node.py" name="serial_node_uno">
	<param name="port" value="/dev/goto_box_usb"/>
	<param name="baud" value="57600"/>
  </node>

  <!-- Run arduino serrial Mega wheelchair control -->
  <node pkg="rosserial_python" type="serial_node.py" name="serial_node_mega">
    <param name="port" value="/dev/wheelchair_mega"/>
    <param name="baud" value="57600"/>
  </node>

  <!-- Run the hokuyo node -->
  <node pkg="hokuyo_node" type="hokuyo_node" name="hokuyo_node" output="screen">
	<param name="port" value="/dev/usb_hokuyo"/>
	<remap from="scan" to="$(arg scan_topic)"/>
  </node>

  <!-- Run tf -->
  <node pkg="tf" type="static_transform_publisher" name="bl_laser" args="-0.23 -0.2 0 0 0 0 /base_link /laser 100" />
  <node pkg="tf" type="static_transform_publisher" name="bf_bl" args="0.0 0 0 0.0 0.0 0.0 /base_link /base_footprint 100" />

  <!-- Run sensing nodes -->
  <node pkg="goto_odroid" type="pos_pros_node_" name="position_processing_node" output="screen"/>
  <node pkg="goto_odroid" type="tf_to_odom_" name="tf_to_odom" output="screen"/>
  <node pkg="goto_odroid" type="loc_" name="loc" output="screen"/>
  <include file="$(find wc_encoders)/vel_to_joy.launch"/>

  <node pkg="wc_encoders" type="wc_encoders_node" name="wc_encoders_node" output="screen"/>

  <!-- Run AMCL -->
  <node pkg="amcl" type="amcl" respawn="true" name="amcl">
	<param name="use_map_topic" value="$(arg use_map_topic)"/>
  	
	<param name="min_particles" value="150"/>
	<param name="max_particles" value="1000"/>
	<param name="kld_err" value="0.05"/> 
	<param name="kld_z" value="0.95"/>
	<param name="update_min_d" value="0.01"/>
	<param name="update_min_a" value="0.01"/>
	<param name="resample_interval" value="1"/>
	<param name="transform_tolerance" value="0.1"/>
	<param name="gui_publish_rate" value="10.0"/>

	<param name="laser_max_beams" value="200"/>
	<param name="laser_z_hit" value="0.95"/>
	<param name="laser_z_short" value="0.01"/>
	<param name="laser_z_max" value="0.05"/>
	<param name="laser_z_rand" value="0.5"/>
	<param name="laser_sigma_hit" value="0.2"/>
	<param name="laser_lambda_short" value="0.1"/>
    <param name="laser_likelihood_max_dist" value="1.0"/>
	<param name="laser_model_type" value="likelihood_field"/>

  	<param name="odom_model_type" value="diff"/>
	<param name="odom_alpha1" value="1.5"/>
	<param name="odom_alpha2" value="1.5"/>
	<param name="odom_alpha3" value="1.5"/>
	<param name="odom_alpha4" value="1.5"/>
	<param name="odom_frame_id" value="odom"/>
	<param name="base_frame_id" value="base_link"/>
	<param name="global_frame_id" value="map"/>
	<param name="tf_broadcast" value="true"/>

	<remap from="scan" to="$(arg scan_topic)"/>
  </node>

  <!-- Run move base -->
  <node pkg="move_base" type="move_base" respawn="true" name="move_base" output="screen">
	<param name="base_global_planner" value="$(arg base_global_planner)"/>
	<param name="base_local_planner" value="$(arg base_local_planner)"/>
	<rosparam file="$(find goto_odroid)/config/planner/base_local_planner.yaml" command="load" />
    	<rosparam file="$(find goto_odroid)/config/planner/costmap_common.yaml" command="load" ns="global_costmap" />
    	<rosparam file="$(find goto_odroid)/config/planner/costmap_common.yaml" command="load" ns="local_costmap" />

	<rosparam file="$(find goto_odroid)/config/planner/local_costmap.yaml" command="load" ns="local_costmap" />
	<param name="local_costmap/width" value="10.0"/>
	<param name="local_costmap/height" value="10.0"/>
    	<rosparam file="$(find goto_odroid)/config/planner/global_costmap.yaml" command="load" ns="global_costmap" unless="$(arg no_static_map)"/>
  </node>

</launch>
